doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,369 @@
+
+- status:  
+String value, the operational port link training status. The list of possible values is as follow  
+
+  | Status      | Description                                                        |
+  |:-----------:|:------------------------------------------------------------------:|
+  | off         | Disabled                                                           |
+  | on          | Enabled, while the further operational status is not yet available |
+  | trained     | Enabled, while the pre-emphasis is tuned successfully              |
+  | not_trained | Enabled, while the pre-emphasis is not yet tuned, no further failure is available |
+  | frame_lock  | Enabled, while the training frame delineation is detected          |
+  | snr_low     | Enabled, while the SNR low threshold is detected                   |
+  | timeout     | Enabled, while the training process is timed out                   |
+
+- supported:  
+String value, the hardware link-training ability status of the switch, this is a constant of the underlying switch silicon regardless of the transceivers attached, the swss#orchagent should populate this information during startup. The list of possible values is as follow  
+
+  | Status      | Description   |
+  |:-----------:|:-------------:|
+  | yes         | Supported     |
+  | no          | Not Supported |
+
+#### SAI attributes
+
+Here is the table to map the fields and SAI attributes:  
+
+| **Parameter**          | **sai_port_attr_t**
+|:-----------------------|:-------------------------------------------|
+| link_training          | SAI_PORT_ATTR_LINK_TRAINING_ENABLE         |
+| link_training_status   | SAI_PORT_ATTR_LINK_TRAINING_RX_STATUS, SAI_PORT_ATTR_LINK_TRAINING_FAILURE_STATUS |
+
+#### YANG Model Enhancements
+
+The port yang model needs to update according to DB schema change. The yang model changes of new fields are described below:
+
+```
+leaf link_training {
+  type string {
+    pattern "on|off";
+  }
+}
+```
+
+These new yang leaf will be added to sonic-port.yang.
+
+#### DB Migration Considerations
+
+By having port link training disabled if "link_training" is not specified in the CONFIG_DB, this feature is fully backward compatible.
+
+#### SWSS Enhancements
+
+##### Port Link-Training Ability Flag
+
+During system startup, PortsOrch should query for the per-port link-training abilities from syncd, and populate these flags into **LINK_TRAINING** table of [STATE_DB](#state-db-enhancements)
+
+##### Port Configuration Flow
+
+The current SONiC port configuration flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+```
+
+The new SONiC port configuration flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if link_training changed and "LT" in xcvr_capabilities:
```

Comment:
Is this relevant since pmon sections are removed?

Reply:
HLD updated